digdag-standards/src/main/java/io/digdag/standards/operator/redshift/RedshiftConnection.java

```diff
@@ -197,26 +197,127 @@ String buildUnloadStatement(UnloadConfig unloadConfig)
-                            + "Or please ask system administrator to prepare a schema that\n"
-                            + "this operator can create tables in and specify `status_table` parameter"
-                            + "to create tables in the schema like `writable_schema.__digdag_status`",
-                            sql);
+            // do nothing
+        }
+
+        @Override
+        public boolean lockedTransaction(UUID queryId, TransactionAction action)
+                throws LockConflictException
+        {
+            beginTransaction();
+            boolean created = createLockedTableWithStatusRow(queryId);
+
+            if (created) {
+                // CREATE TABLE AS succeeded in this transaction.
+                // No other threads / nodes has succeeded to commit before.
+                // This transaction is responsible to run the action.
+                action.run();
+                updateStatusRowAndCommitTransaction(queryId);
+                return true;
+            }
+            else {
+                // CREATE TABLE AS conflicted.
+                // A thread / node succeeded to commit before.
+                // Skip the action.
+                abortTransaction();
+                return false;
+            }
+        }
+
+        private void beginTransaction()
+        {
+            executeStatement("begin a transaction", "BEGIN");
+        }
+
+        private void abortTransaction()
+        {
+            executeStatement("rollback a transaction", "ROLLBACK");
+        }
+
+        private void updateStatusRowAndCommitTransaction(UUID queryId)
+        {
+            executeStatement("update status row",
+                    String.format(ENGLISH,
+                        "UPDATE %s SET completed_at = CURRENT_TIMESTAMP WHERE query_id = '%s'",
+                        escapeIdent(statusTableName(queryId)),
+                        queryId.toString())
+                    );
+            executeStatement("commit updated status row", "COMMIT");
+        }
+
+        private void executeStatement(String desc, String sql)
+        {
+            try {
+                execute(sql);
+            }
+            catch (SQLException ex) {
+                throw new DatabaseException("Failed to " + desc, ex);
+            }
+        }
+
+        private boolean createLockedTableWithStatusRow(UUID queryId)
+        {
+            String sql = String.format(ENGLISH,
+                    "CREATE TABLE IF NOT EXISTS %s" +
+                    " (query_id, created_at, completed_at)" +
+                    " AS SELECT '%s'::text, CURRENT_TIMESTAMP::timestamptz, NULL::timestamptz",
+                    escapeIdent(statusTableName(queryId)),
+                    queryId.toString());
+            try {
+                execute(sql);
+                return true;
+            }
+            catch (SQLException ex) {
+                if (isConflictException(ex)) {
```

Comment:
Using the existence of a status table itself as something like a lock would work to avoid duplicated adding records, but it sounds that retries will keep failing if the first connection is disconnected after creating a status table...